### ISSUE TITLE
App Check debug provider(s) provide a difficult developer experience

### ISSUE THREAD

**Issue by mikehardy** [firebase/firebase-ios-sdk#8584]
### Step 0: Are you in the right place?

  * For issues or feature requests related to __the code in this repository__
    file a Github issue.
    * If this is a __feature request__ please use the Feature Request template.
  * For general technical questions, post a question on [StackOverflow](http://stackoverflow.com/)
    with the `firebase` tag.
  * For general (non-iOS) Firebase discussion, use the [firebase-talk](https://groups.google.com/forum/#!forum/firebase-talk)
    google group.
  * For backend issues, console issues, and other non-SDK help that does not fall under one
    of the above categories, reach out to
    [Firebase Support](https://firebase.google.com/support/).
  * Once you've read this section and determined that your issue is appropriate for
    this repository, please delete this section.

### [REQUIRED] Step 1: Describe your environment

  * Xcode version: 12.5.1
  * Firebase SDK version: 8.6.0
  * Installation method: `CocoaPods`
  * Firebase Component: App Check

#### Steps to reproduce:

This is perhaps more of a feature request, but it manifests in lots of bugs. [I was requested](https://github.com/FirebaseExtended/flutterfire/issues/6794#issuecomment-904813920) by @maksymmalyhin to post a separate issue explaining the difficulties so here we are!

What happened? How can we make the problem occur?

In this comment I laid out the basic difficulties, it's important to understand them, it should be considered required reading for the issue but I don't want to duplicate it: https://github.com/FirebaseExtended/flutterfire/issues/6794#issuecomment-904796629

It has lots of specific references to code that show the difficulty, it really is a solid read I think. Take your time :-)

#### Relevant Code:

Permalinks in the comment referenced above

## Proposed Solution

My thoughts on a solution involve two key changes to the App Check system, that should span Web + Apple + Android:

### Allow App Check activate with Provider changes after Firebase App Configure  on all platforms

This is critical for an easy developer experience as interpreted language ecosystems (FlutterFire, react-native-firebase, others) cannot easily configure things at the native level (in `AppDelegate.m` prior to native configure) via AppDelegate (or java Application) native code.

If you can change the App Check Provider at runtime, then this is fine - an interpreted language framework that wraps App Check can simply determine (in whatever way is best for them!) that they are in debug mode or should use App Attest or similar post-FirebaseApp.configure and switch their Provider.

I think it's fine to put a constraint that the AppCheck Provider may not change after using any other services that App Check works with (similar to the useEmulator API usage constraints), although it would be best if the Provider could change any time and it simply purged existing tokens.

### Alter DebugProviders APIs to allow configuration of shared secret via parameter

This is similar to the problem above but separate. Developers in interpreted language ecosystems frequently use configuration parameter injection frameworks that work at the dynamic language layer, and that's not running until well after the native layer is configured. When they determine they are running in an environment where they want to use the Debug Providers, they should be able to instantiate the DebugProvider with a shared secret from the dynamic language layer, then activate it.

The combination of these two changes across the various SDKs would make App Check adoption much easier for the huge audiences served via FlutterFire and react-native-firebase

Thanks!

**Comment by ReenaSuoliman** [comment]
Thank you

**Comment by ReenaSuoliman** [comment]
https://github.com/firebase/firebase-ios-sdk/issues/8584#issue-981408250

**Comment by maksymmalyhin** [comment]
Thank you for the feedback and proposed solutions! We will consider it and alternatives internally and post updates here.

**Comment by maksymmalyhin** [comment]
Internally tracked in b/197661673.

**Comment by maksymmalyhin** [comment]
cc @Feiyang1 @rosalyntan

**Comment by mikehardy** [comment]
Thanks - and as you mention "alternatives internally" there are always lots of ways to do things, I'll be listening here and my github email is also valid in case you all come up with something totally different and want a sounding board. Cheers

**Comment by paulb777** [comment]
@mikehardy While I don't believe that this specific fr has been addressed, several App Check usability improvements have been made in the last few years. Is this still a need?

**Comment by mikehardy** [comment]
We have a reasonable stable "let's paper over it" solution over at react-native-firebase. Can't speak to FlutterFire repo

Our solution was to use the firebase-js-sdk APIs to work around the static-initialization-on-startup thinking baked in to the native App Check implementations and I was able to

Parts are:
- javascript API as basis (our foundational API, as react-native-firebase - which provides javascript bindings for the native SDKs is designed to be a drop-in replacement for firebase-js-sdk)
- App Check custom provider API as specific loophole (like a network localhost proxy...)
- ReactNative-specific custom provider type that provides ability to do a runtime configuration post-startup (so javascript runtime in react-native is available...)
- Native custom provider that does a facade over the built-in native providers here (the target of the proxy)

It's a lot of shimming and facading but...in the end it works and I moved on :-)

It would be better if the native stuff didn't make assumptions about being configured once and only once on startup though, that was the key hindrance

https://github.com/invertase/react-native-firebase/pull/6847
